zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs

```diff
@@ -0,0 +1,229 @@
+        cb.stack_pop(bytes_data_offset.expr());
+        cb.stack_pop(bytes_length.expr());
+
+        // Transition to internal state CopyToMemory
+        let next_src_addr = cb.query_cell_next_step();
+        let next_dst_addr = cb.query_cell_next_step();
+        let next_bytes_left = cb.query_cell_next_step();
+        let next_src_addr_end = cb.query_cell_next_step();
+        let next_from_tx = cb.query_cell_next_step();
+        let next_tx_id = cb.query_cell_next_step();
+        cb.require_next_state(ExecutionState::CopyToMemory);
+        cb.add_constraint(
+            "next_src_addr = data_offset",
+            next_src_addr.expr() - data_offset.clone(),
+        );
+        cb.add_constraint(
+            "next_dst_addr = memory_offset",
+            next_dst_addr.expr() - memory_offset,
+        );
+        cb.add_constraint(
+            "next_bytes_left = length",
+            next_bytes_left.expr() - length,
+        );
+        cb.add_constraint(
+            "next_src_addr_end = data_offset + calldata_length",
+            next_src_addr_end.expr() - data_offset - calldata_length.expr(),
+        );
+        cb.add_constraint(
+            "next_from_tx = is_root",
+            next_from_tx.expr() - cb.curr.state.is_root.expr(),
+        );
+        cb.add_constraint(
+            "next_tx_id = tx_id",
+            next_tx_id.expr() - tx_id.expr(),
+        );
+
+        // State transition
+        let step_state_transition = StepStateTransition {
+            // We keep the program_counter same because Calldatacopy hasn't
+            // finished the copy yet
+            rw_counter: Delta(3.expr()), // 3 stack pop
+            //program_counter: Delta(1.expr()),
+            stack_pointer: Delta(3.expr()),
+            memory_size: To(memory_expansion.next_memory_size()),
+            ..Default::default()
+        };
+        let same_context = SameContextGadget::construct(
+            cb,
+            opcode,
+            step_state_transition,
+            Some(memory_expansion.gas_cost()),
+        );
+
+        Self {
+            same_context,
+            memory_offset: bytes_memory_offset,
+            data_offset: bytes_data_offset,
+            length: bytes_length,
+            tx_id,
+            calldata_length,
+            memory_expansion,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        block: &Block<F>,
+        tx: &Transaction<F>,
+        _: &Call<F>,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let [memory_offset, data_offset, length] =
+            [step.rw_indices[0], step.rw_indices[1], step.rw_indices[2]]
+            .map(|idx| block.rws[idx].stack_value());
+        self.memory_offset.assign(
+            region,
+            offset,
+            Some(memory_offset.to_le_bytes()[..5].try_into().unwrap()),
```

Comment:
suggest use `..NUM_ADDRESS_BYTES_USED` here, like in https://github.com/appliedzkp/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/execution/memory.rs#L168

Reply:
fixed